ENH: add mm->qm divmod

tylerjereddy opened the pull request:
~This still needs work -- something isn't quite right according to the unit tests.~

Would be nice if an inner loop that performs the operations of two other inner loops could just gracefully call the those other two inner loops, but I think Eric & I have had that discussion before & there are issues with i.e., nesting binop loops.

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -1644,6 +1644,41 @@ TIMEDELTA_mm_q_floor_divide(char **args, npy_intp *dimensions, npy_intp *steps,
+        }
+        else if (in2 == 0) {
+            npy_set_floatstatus_divbyzero();
+            *((npy_timedelta *)op1) = 0;
```
Isn't `op1` an `int64` rather than a `timedelta`? I know the cast is the same either way, but it would be more readable to have these types right.

tylerjereddy replied:
You asked me to do the reverse in the recent floordiv PR, but I've just gone ahead and changed as requested here since there's really no effect as far as I know.

eric-wieser replied:
Did I?

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -1644,6 +1644,41 @@ TIMEDELTA_mm_q_floor_divide(char **args, npy_intp *dimensions, npy_intp *steps,
+        else if (in2 == 0) {
+            npy_set_floatstatus_divbyzero();
+            *((npy_timedelta *)op1) = 0;
+            *((npy_timedelta *)op2) = 0;
```
Based on how `divmod(np.array(1.0), np.array(0))` behaves, this should be `NAT`.

tylerjereddy replied:
I'm not sure I understand -- I haven't added any new code here -- it just the `m8` remainder and floordiv code paths pasted into a new inner loop together.

If we do this, the modulus operator would produce a different result when used in isolation vs. in the context of divmod.

eric-wieser replied:
Hmm, I didn't check the modulus code. You're right that what you have here is consistent with modulus.

On the other hand, I think we might have gotten modulus wrong. If we're going to change modulus on timedeltas, it might be better to do so before introducing divmod, so that we change behavior in one place and not two,

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -1644,6 +1644,41 @@ TIMEDELTA_mm_q_floor_divide(char **args, npy_intp *dimensions, npy_intp *steps,
+                *((npy_timedelta *)op2) = rem + in2;
+            }
+
+            if (((in1 > 0) != (in2 > 0)) && (in1 % in2 != 0)) {
```
This condition is the same as the one above - can you combine the ifs?

tylerjereddy replied:
ok, I've simplified the conditionals

eric-wieser commented on numpy/core/tests/test_datetime.py:
```diff
@@ -1161,6 +1161,33 @@ def test_timedelta_floor_div_error(self, val1, val2):
+        ])
+    def test_timedelta_divmod(self, op1, op2):
+        expected = (op1 // op2, op1 % op2)
+        assert_equal(divmod(op1, op1), expected)
```
Would be good to have some tests for divide-by-zero / NaT too

tylerjereddy replied:
Yup, would also be good if I didn't have `op1` as both operands in the `divmod` -- no wonder the tests were failing!

tylerjereddy commented:
Revised based on feedback so far -- if `m8` remainder operation really needs a fix I suppose we could apply that here as well rather than having `divmod` behave differently.

eric-wieser commented:
We might want to consult the mailing list on whether we should revise the behavior of `timedelta64(1) % timedelta64(0)`. Whatever we decide, obviously `divmod` should match.

Otherwise looks good to me.

eric-wieser commented:
Ah, I'd forgotten how recently we added modulo support (#12120), so we can probably get away with fixing it, especially if we backport. Was there any discussion about what %0 should return there?

miccoli commented:
@eric-wieser : I do not remember that in #12120 the %0 case was discussed: I would suggest to consider the current behaviour (`1 % 0 == 0`)  a bug.

There are still some open points in the  `np.timedelta64` implementation which were discussed in #12120:
1. `np.timedelta64(13) / 7` vs `np.timedelta64(13) //  7`
1. `np.timedelta64(13) / np.timedelta(7)` vs `np.timedelta64(13) // np.timedelta(7)`

IMHO currently (1.16.0rc2) we have
- correct: 
  - `np.timedelta64(13) //  7 ==  np.timedelta64(1)`
  - `np.timedelta64(13) / np.timedelta64(7) == 1.857_142_857_142_857_2`
- wrong: 
  - `np.timedelta64(13) / 7 == np.timedelta64(1)` 
     which instead should be 
    `np.timedelta64(2)` or 
    `np.timedelta64(1_857, 'ms')` or 
    `np.timedelta64(1_857_143, 'us')` or
    `numpy.timedelta64(1_857_142_857,'ns')` depending on how (or whether) dividend and divisor are promoted before division

- not implemented:
  - `np.timedelta64(13) // np.timedelta64(7)` 
      which should be `1` according to the invariant
      `a == (a // b)*b + a%b`

tylerjereddy commented:
@miccoli 

-  floordiv with two `timedelta64` operands should be available in master now
- I didn't touch the cases involving `timedelta64` with integer operands recently, so those have probably been that way for a while -- not sure how disruptive those behavior changes would be to make now / try to backport
- looks like both you / mailing list agree that the modulus % 0 case for two `timedelta64` operands is a bug, and should return `NaT`; I'll probably aim to have one commit that fixes that bug in remainder (to facilitate possible backport), and then a second commit to add divmod with the revised remainder behavior baked in

tylerjereddy committed:
BUG: timedelta64 % 0 behavior

* modulus operation with two timedelta64
operands now returns NaT in the case of
division by zero, rather than returning
zero

tylerjereddy committed:
ENH: add mm->qm divmod

* add timedelta64 divmod inner
loop with type signature mm->qm

* also adjusted typecasting in
mm_q floor division to match the
approach used here (should not
change any behavior)

tylerjereddy commented:
Revised the PR with separate commits for the remainder bug fix and the feature addition / typecast cleanup, so that the former may be backported if desired.

eric-wieser commented:
I'd probably be in favor of just backporting everything

eric-wieser commented on numpy/core/tests/test_datetime.py:
```diff
@@ -1161,6 +1161,52 @@ def test_timedelta_floor_div_error(self, val1, val2):
+        with assert_warns(RuntimeWarning):
+            actual = divmod(op1, op2)
+            expected = (op1 // op2, op1 % op2)
+            assert_equal(actual, expected)
```
I think this would be better as:

```python
with assert_warns(RuntimeWarning):
    expected = (op1 // op2, op1 % op2)
with assert_warns(RuntimeWarning):
    actual = divmod(op1, op2)
assert_equal(actual, expected)
```

Assuming you want to show that both paths warn.

eric-wieser reviewed (approved):
Looks good to me, but needs a release note. Since this is a simple extension, and `divmod` is something I'd expect to exist alongside `%`, I think we should all of this in 1.16, not just the bug fix.

Needs a release note in either 1.16 or 1.17, depending on what we decide.

tylerjereddy commented:
Refined the test as requested & removed WIP tag.

I'll hold off on the release note until we have a decision on the milestone target.

seberg commented:
Considering how super new it is, I guess it is fine to put the NaN fix definitely into 1.16 (even 1.16.1, don't want to make things weird with already rc's out for long). Backporting the new functionality seems OK to me as well, since even if there is something wrong, I don't think it can possibly cause a regression. If this wasn't the LTS I would probably be against it.

So I would say: lets just target 1.16.1.

charris commented:
If we want it in 1.16, should try rebasing on that branch with a release note, then forward port without the release note. The 1.16 release notes will be forward ported after release. I would like to make the 1.16 release this weekend.

tylerjereddy commented:
I'll tentatively add the `1.16.1` milestone & try the rebase thing locally first, etc. I've been stuck working on the SciPy paper revisions / organization for most of today.

eric-wieser commented:
Rebasing on f07a38da97a6a36eb12b203f6c1ffa4bf2b2cb87 will make both the forward and backport easy

charris commented:
Needs a release note. Put it in the 1.17.0-notes and I'll copy it to the 1.16.1-notes. @eric-wieser backports are really easy anyway unless there are a ton of commits with conflicts. I'll be interested to see if Tyler can still add the release note without problems.

tylerjereddy commented:
Ok, added a release note, rebased on latest master, and force pushed. This will obviously increase the likelihood of master branch merge conflict eventually.

eric-wieser commented on numpy/core/tests/test_datetime.py:
```diff
@@ -1161,6 +1161,53 @@ def test_timedelta_floor_div_error(self, val1, val2):
+        assert_equal(divmod(op1, op2), expected)
+
+    @pytest.mark.parametrize("op1, op2", [
+        # reuse cases from floordiv
```
Can you reuse them in the source by creating a global variable with these cases?

mattip commented:
two small nits, but not critical. Will merge later today

mattip commented:
Thanks Tyler

charris commented:
I've made the additions to the 1.16.1 release notes.
